cmd/shader_compiler/src/header_generator.cpp

```diff
@@ -0,0 +1,232 @@
+#include "header_generator.hpp"
+
+#include <components/common/logging.hpp>
+
+#include <fstream>
+#include <iostream>
+#include <string>
+
+namespace shader_compiler
+{
+using namespace components;
+
+namespace detail
+{
+std::string snake_to_pascal(const std::string &snake)
+{
+	std::string pascal;
+
+	for (size_t i = 0; i < snake.size(); i++)
+	{
+		if (i == 0)
+		{
+			pascal += toupper(snake[0]);
+			continue;
+		}
+
+		if (snake[i] == '_')
+		{
+			pascal += toupper(snake[i + 1]);
+			i++;
+			continue;
+		}
+
+		pascal += snake[i];
+	}
+
+	return pascal;
+}
+}        // namespace detail
+
+void generate_raw_variant_pointers(const ShaderDescriptor &descriptor, std::ofstream &output_stream)
+{
+	output_stream << "// clang-format off" << std::endl;
+
+	for (const auto &variant_pair : descriptor.variants)
+	{
+		auto &hash    = variant_pair.first;
+		auto &variant = variant_pair.second;
+
+		output_stream << "\tconstexpr uint32_t " << detail::snake_to_pascal(descriptor.name) << "__" << hash << "[] = [" << std::endl;
```

Comment:
As the function `get_variant` can't be `constexpr`, you could as well make this a `const std::vector<uint32_t>`, and let `get_variant` return a `const std::vector<uint32_t> &`, without any copy operations.